misk/src/main/kotlin/misk/flags/DynamicFlagModule.kt

```diff
@@ -0,0 +1,163 @@
+                .asEagerSingleton()
+    }
+
+    fun bindIntFlag(name: String, desc: String, qualifier: Annotation = Names.named(name)) {
+        bind<IntFlag>()
+                .annotatedWith(qualifier)
+                .toProvider(IntFlagProvider(name, desc))
+                .asEagerSingleton()
+    }
+
+    fun bindDoubleFlag(name: String, desc: String, qualifier: Annotation = Names.named(name)) {
+        bind<DoubleFlag>()
+                .annotatedWith(qualifier)
+                .toProvider(DoubleFlagProvider(name, desc))
+                .asEagerSingleton()
+    }
+
+    inline fun <reified T : Any> bindJsonFlag(
+            name: String,
+            desc: String,
+            qualifier: Annotation = Names.named(name)
+    ) = bindJsonFlag(T::class, name, desc, qualifier)
+
+    fun <T : Any> bindJsonFlag(
+            kclass: KClass<T>,
+            name: String,
+            desc: String,
+            a: Annotation = Names.named(name)
+    ) {
+        val flagType = parameterizedType<JsonFlag<T>>(kclass.java)
+
+        @Suppress("UNCHECKED_CAST")
+        val flagTypeLiteral = TypeLiteral.get(flagType) as TypeLiteral<JsonFlag<T>>
+        bind(flagTypeLiteral)
+                .annotatedWith(a)
+                .toProvider(JsonFlagProvider(kclass, name, desc))
+                .asEagerSingleton()
+    }
+
+    private class BooleanFlagProvider(val name: String, val desc: String) : Provider<BooleanFlag> {
+        @Inject
+        lateinit var flagStore: DynamicFlagStore
+
+        override fun get(): BooleanFlag = flagStore.registerBooleanFlag(name, desc)
+    }
+
+    private class StringFlagProvider(val name: String, val desc: String) : Provider<StringFlag> {
+        @Inject
+        lateinit var flagStore: DynamicFlagStore
+
+        override fun get(): StringFlag = flagStore.registerStringFlag(name, desc)
+    }
+
+    private class IntFlagProvider(val name: String, val desc: String) : Provider<IntFlag> {
+        @Inject
+        lateinit var flagStore: DynamicFlagStore
+
+        override fun get(): IntFlag = flagStore.registerIntFlag(name, desc)
+    }
+
+    private class DoubleFlagProvider(val name: String, val desc: String) : Provider<DoubleFlag> {
+        @Inject
+        lateinit var flagStore: DynamicFlagStore
+
+        override fun get(): DoubleFlag = flagStore.registerDoubleFlag(name, desc)
+    }
+
+    private class JsonFlagProvider<T : Any>(
+            val kclass: KClass<T>,
+            val name: String,
+            val desc: String
+    ) : Provider<JsonFlag<T>> {
+        @Inject
+        lateinit var flagStore: DynamicFlagStore
+
+        @Inject
+        lateinit var moshi: Moshi
+
+        override fun get(): JsonFlag<T> {
+            val stringFlag = flagStore.registerStringFlag(name, desc)
+            val adapter = moshi.adapter(kclass.java)
+            return JsonFlag(stringFlag, adapter)
+        }
+    }
+
+    private class FlagsProvider<T : Any>(
```

Comment:
Would it be simpler/safer to just have the user provide a function rather than trying to convert the constructor into one?